src/rpc/rpc_payment.cpp

```diff
@@ -0,0 +1,389 @@
+
+    m_hashrate[now] += m_diff;
+    add64clamp(m_credits_total, m_credits_per_hash_found);
+    add64clamp(info.credits_total, m_credits_per_hash_found);
+    ++m_nonces_good;
+    ++info.nonces_good;
+
+    credits = info.credits;
+    block = info.block;
+    block.nonce = nonce;
+    stale = !is_current;
+    return true;
+  }
+
+  bool rpc_payment::foreach(const std::function<bool(const crypto::public_key &client, const client_info &info)> &f)
+  {
+    for (std::unordered_map<crypto::public_key, client_info>::const_iterator i = m_client_info.begin(); i != m_client_info.end(); ++i)
+    {
+      if (!f(i->first, i->second))
+        return false;
+    }
+    return true;
+  }
+
+  bool rpc_payment::load(const std::string &directory)
+  {
+    TRY_ENTRY();
+    m_directory = directory;
+    std::string state_file_path = directory + "/" + RPC_PAYMENTS_DATA_FILENAME;
+    MINFO("loading rpc payments data from " << state_file_path);
+    std::ifstream data;
+    data.open(state_file_path, std::ios_base::binary | std::ios_base::in);
+    if (!data.fail())
+    {
+      try
+      {
+        boost::archive::portable_binary_iarchive a(data);
+        a >> *this;
+      }
+      catch (const std::exception &e)
+      {
+        MERROR("Failed to load RPC payments file: " << e.what());
+        m_client_info.clear();
+      }
+    }
+    else
+    {
+      m_client_info.clear();
+    }
+
+    CATCH_ENTRY_L0("rpc_payment::load", false);
+    return true;
+  }
+
+  bool rpc_payment::store(const std::string &directory_)
+  {
+    TRY_ENTRY();
+    const std::string &directory = directory_.empty() ? m_directory : directory_;
+    MDEBUG("storing rpc payments data to " << directory);
+    if (!tools::create_directories_if_necessary(directory))
+    {
+      MWARNING("Failed to create data directory: " << directory);
+      return false;
+    }
+    std::string state_file_path = (boost::filesystem::path(directory) / RPC_PAYMENTS_DATA_FILENAME).string();
+    if (epee::file_io_utils::is_file_exist(state_file_path))
+    {
+      std::string state_file_path_old = state_file_path + ".old";
+      boost::system::error_code ec;
+      boost::filesystem::remove(state_file_path_old, ec);
+      std::error_code e = tools::replace_file(state_file_path, state_file_path_old);
+      if (e)
+        MWARNING("Failed to rename " << state_file_path << " to " << state_file_path_old << ": " << e);
+    }
+    std::ofstream data;
+    data.open(state_file_path, std::ios_base::binary | std::ios_base::out | std::ios::trunc);
+    if (data.fail())
+    {
+      MWARNING("Failed to save RPC payments to file " << state_file_path);
+      return false;
+    };
+    boost::archive::portable_binary_oarchive a(data);
+    a << *this;
+    return true;
+    CATCH_ENTRY_L0("rpc_payment::store", false);
+    return true;
+  }
+
+  unsigned int rpc_payment::flush_by_age(time_t seconds)
+  {
+    unsigned int count = 0;
+    const time_t now = time(NULL);
```

Comment:
man 2 time says it's seconds since the epoch. I never heard of time being anything else than seconds before. That change seems busywork to me.

Reply:
The man page is referring specifically to the GNU implementation. Seconds is only mandated by Posix, not ISO C. From section 7.27 of the last [C11 draft](http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1570.pdf) (before you gotta pay):

> The range and precision of times representable in `clock_t` and `time_t` are implementation-defined.

Definitely surprising - its why working with time in a purely portable manner in C makes me want to bash something. Luckily, I doubt we'll ever use an implementation that is _not_ seconds since 1970 (just varied range/integer types).